src/uart.rs

```diff
@@ -0,0 +1,355 @@
+//! Access to [RIOT's UART](https://doc.riot-os.org/group__drivers__periph__uart.html)
+//!
+//! Author: Kilian Barning <barning@uni-bremen.de>
+
+use core::ptr;
+
+use crate::error::{NegativeErrorExt, NumericError};
+use riot_sys::libc::{c_uint, c_void};
+use riot_sys::*;
+
+/// This enum representatives the status returned by various `UART`-functions
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum UartDeviceError {
+    InvalidDevice,
+    UnsupportedConfig,
+    Other,
+}
+
+impl UartDeviceError {
+    /// Converts the given `c_int` into the matching Enum representation
+    fn from_c(n: NumericError) -> Self {
+        match n {
+            crate::error::ENODEV => Self::InvalidDevice,
+            crate::error::ENOTSUP => Self::UnsupportedConfig,
+            _ => Self::Other,
+        }
+    }
+}
+
+#[cfg(riot_module_periph_uart_modecfg)]
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum DataBits {
+    Five,
+    Six,
+    Seven,
+    Eight,
+}
+
+#[cfg(riot_module_periph_uart_modecfg)]
+impl DataBits {
+    fn to_c(self) -> uart_data_bits_t {
+        match self {
+            Self::Five => uart_data_bits_t_UART_DATA_BITS_5,
+            Self::Six => uart_data_bits_t_UART_DATA_BITS_6,
+            Self::Seven => uart_data_bits_t_UART_DATA_BITS_7,
+            Self::Eight => uart_data_bits_t_UART_DATA_BITS_8,
+        }
+    }
+}
+
+#[cfg(riot_module_periph_uart_modecfg)]
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum Parity {
+    None,
+    Even,
+    Odd,
+    Mark,
+    Space,
+}
+
+#[cfg(riot_module_periph_uart_modecfg)]
+impl Parity {
+    fn to_c(self) -> uart_parity_t {
+        match self {
+            Self::None => uart_parity_t_UART_PARITY_NONE,
+            Self::Even => uart_parity_t_UART_PARITY_EVEN,
+            Self::Odd => uart_parity_t_UART_PARITY_ODD,
+            Self::Mark => uart_parity_t_UART_PARITY_MARK,
+            Self::Space => uart_parity_t_UART_PARITY_SPACE,
+        }
+    }
+}
+
+#[cfg(riot_module_periph_uart_modecfg)]
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum StopBits {
+    One,
+    Two,
+}
+
+#[cfg(riot_module_periph_uart_modecfg)]
+impl StopBits {
+    fn to_c(self) -> uart_stop_bits_t {
+        match self {
+            Self::One => uart_stop_bits_t_UART_STOP_BITS_1,
+            Self::Two => uart_stop_bits_t_UART_STOP_BITS_2,
+        }
+    }
+}
+
+/// This struct contains the `UART` device and handles all operation regarding it
+///
+/// The lifetime `'cb` indicates how long a registered callback on the device lives; for many
+/// cases, that is `'static`.
+///
+/// [UART implementation]: https://doc.riot-os.org/group__drivers__periph__uart.html
+#[derive(Debug)]
+pub struct UartDevice<'cb> {
+    dev: uart_t,
+    _phantom: core::marker::PhantomData<&'cb ()>,
+}
+
+impl<'cb> UartDevice<'cb> {
```

Comment:
Why is this block separated from the identical `impl<'cb> UartDevice<'cb>` in line 246?

Reply:
Because I kept the order of functions rather than keeping impls together. In other words historical reasons, hysterical raisins, no reason at all.

Shuffling them around (along with others for better documentation structure); not sure where to best put that during squashing.